Gutenberg: load blocks from CDN

simison opened the pull request:
A suggestion for easy testing of Gutenberg Blocks in Jetpack.

This PR doesn't explore how to _build_ these blocks, but how to _load_ them in Jetpack in one file.

This should be flexible enough to allow very easy testing as well as easier local development by allowing to disable CDN.

Note that this contains only `editor` side assets.

#### Testing instructions:
1. Apply D16998-code to your sandbox to get `https://s0.wp.com/wp-content/mu-plugins/jetpack/_inc/blocks/jetpack-editor.js`
2. Confirm that by default nothing happens in Gutenberg editor
3. Confirm that you can enable these blocks via filter:
    ```php
    add_filter( 'jetpack_gutenberg', '__return_true', 10 );
    ```
    (You can place this e.g. in a file at `mu-plugins` -folder)
4. Confirm that by default files are loaded via CDN, but that you can disable this via filter:
    ```php
    add_filter( 'jetpack_gutenberg_cdn', '__return_false', 10 );
    ```
    Note that you would need to build files manually by running this in Calypso:
    ```
    npm run sdk:gutenberg -- \
	    --editor-script=client/gutenberg/extensions/presets/jetpack/editor.js \
	    --output-dir=~/jetpack/_inc/blocks
    ```
5. Confirm that by default CDN has a date +Jetpack version as a cachebuster, but that you can change it by filter:
    ```php
    add_filter( 'jetpack_gutenberg_cdn_cache_buster', function( $version ) { return time(); }, 10, 1 );
    ```

Blocks that you should have in Gutenberg editor:
- Tiled gallery
- Markdown
- Publicize

## TODO

- I had issues getting blocks working but it might be due to some changes in blocks and/or Gutenberg — I need to have a closer look (ping @tyxla)
- CSS file doesn't exist yet, it's waiting for https://github.com/Automattic/wp-calypso/pull/26551

zinigor commented on class.jetpack.php:
```diff
@@ -526,6 +526,14 @@ private function __construct() {
+		add_action( 'init', array( $this, 'load_jetpack_gutenberg' ) );
+
+		// A filter to control Gutenberg editor blocks
+		add_filter( 'jetpack_gutenberg', '__return_false', 9);
```
Please add a space between `9` and the closing brace.

simison commented:
Updated CDN assets (D16998-code) with the latest versions:
- removed publicize
- contains styles, too

If you type `npm link` in your Calypso-folder, you can then just run this in your local sandbox folder to get the files:

```
calypso-sdk gutenberg --editor-script=client/gutenberg/extensions/presets/jetpack/editor.js --output-dir=./wp-content/mu-plugins/jetpack/_inc/blocks
```

oskosk commented on class.jetpack.php:
```diff
@@ -7258,4 +7261,66 @@ public static function handle_post_authorization_actions( $activate_sso = false,
+			$version
+		);
+
+		if ( function_exists( 'register_block_type' ) ) {
```
Can we have a method like `Jetpack::is_gutenberg_available` or such that wraps this check that probably will get used a lot ?

oskosk commented on class.jetpack.php:
```diff
@@ -7258,4 +7261,66 @@ public static function handle_post_authorization_actions( $activate_sso = false,
+	 *   add_filter( 'jetpack_gutenberg_cdn_cache_buster', function( $version ) { return time(); }, 10, 1 );
+	 */
+	public static function load_jetpack_gutenberg() {
+		if ( ! apply_filters( 'jetpack_gutenberg', false ) || ! Jetpack::is_plugin_active( 'gutenberg/gutenberg.php' ) ) {
```
this check `Jetpack::is_plugin_active( 'gutenberg/gutenberg.php' ) )` will break if gutenberg is not there as a plugin (as will happen on 5.0 and probably nighlty releases of WP 5.0 as soon as somebody commits gutenberg into it(. 

Would it be enough to just check for the availability of `register_block_type` or , as mentioned in my next comment, using a `Jetpack::is_gutenberg_available` handy function ?

simison commented:
Updated to use `is_gutenberg_available()` as per @oskosk suggestion and added some inline docs.

ockham commented on class.jetpack.php:
```diff
@@ -7258,4 +7261,107 @@ public static function handle_post_authorization_actions( $activate_sso = false,
+		 */
+		if ( apply_filters( 'jetpack_gutenberg_cdn', true ) ) {
+			$editor_script = 'https://s0.wp.com/wp-content/mu-plugins/jetpack/_inc/blocks/jetpack-editor.js';
+			$editor_style = 'https://s0.wp.com/wp-content/mu-plugins/jetpack/_inc/blocks/jetpack-editor.css';
```
Does JP have any way of rotating through `s{0,1,2}`?

ockham replied:
(Or more generally, to conditionally load code from the CDN or locally, respectively -- much like we're doing here?)

simison replied:
I kept things simple for this since its primary use will be for development and testing only.

The closest thing I found was in the Masterbar but in many places around Jetpack we just load via `s0`:

https://github.com/Automattic/jetpack/blob/1f532fb9e5d781c8f55c5669f025dea2453b72c3/modules/masterbar/masterbar.php#L146-L156

We have also have nice stuff like these helpers:

```php
Jetpack::dns_prefetch( array(
	'//s0.wp.com',
	'//s1.wp.com',
	'//s2.wp.com',
) );
```
and
https://github.com/Automattic/jetpack/blob/6a5b8de344946c99ef25cea17e119628acedb27f/class.jetpack.php#L7180-L7199

simison replied:
>Or more generally, to conditionally load code from the CDN or locally, respectively -- much like we're doing here

Even if there was, I needed to be able to control these via filters so that they wouldn't load even when in development mode.
